### PR TITLE
Fall back to the QR solver when the Cholesky factorization in lstsq fails

### LINKED ISSUE #115505: tf.linalg.lstsq returns NaN for rank-deficient matrices despite well-defined least-norm solution
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.11

### Custom code

Yes

### OS platform and distribution

Linux (Ubuntu 22.04)

### Mobile device

_No response_

### Python version

Python 3.10

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

tf.linalg.lstsq returns NaN values when applied to a rank-deficient (singular) Gram matrix, even though a well-defined least-norm solution exists.

Specifically, when solving a least-squares problem with a matrix A = X @ X^T where X has linearly dependent rows (rank-deficient), the solver produces NaN outputs instead of a valid least-norm solution.

This suggests that the current implementation does not robustly handle ill-conditioned or singular systems, possibly due to missing singular value thresholding or lack of regularization during decomposition (e.g., SVD or QR).

Expected behavior:
The function should return a finite least-norm solution (e.g., via Moore–Penrose pseudoinverse), or at minimum provide a warning/error indicating numerical instability, rather than silently producing NaN values.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

# Construct a rank-deficient matrix (all rows identical → rank = 1)
x = tf.constant([
    [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
], dtype=tf.float32)

# Gram matrix (rank-deficient)
gram = tf.matmul(x, x, transpose_b=True)

# RHS
rhs = tf.ones_like(gram)

# Solve least squares
sol = tf.linalg.lstsq(gram, rhs)

# Check for NaN
has_nan = tf.reduce_any(tf.math.is_nan(sol))

print("Solution:")
print(sol.numpy())
print("Contains NaN:", bool(has_nan))
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by Phoebus-Liu**
`tf.linalg.lstsq` returns NaN for a singular matrix, even though a finite
minimum-norm solution exists. Reported in #115505.

With `fast=True` (the default) and a matrix small enough to use the composite
implementation, the solution is computed from a Cholesky factorization of the
Gramian (`_RegularizedGramianCholesky`). That factorization does not exist when
the matrix is singular; the factorization then fails and its output is filled
with NaN, which propagates to the result:

    x = [[1, 0, ..., 0]] * 4
    gram = x @ x.T                              # rank 1, singular
    tf.linalg.lstsq(gram, tf.ones_like(gram))   # -> all NaN

The QR based kernel (`fast=False`) does handle these systems and returns the
minimum-norm solution, so this change falls back to it when the Cholesky
factorization failed. The fallback is taken only when the composite result
contains NaN, which is the observable symptom of the failed factorization.
`is_nan` is not defined for complex types, so for those we check the real and
imaginary parts separately via `math_ops.is_nan(math_ops.real(...))` /
`math_ops.is_nan(math_ops.imag(...))` combined with `logical_or`, instead of
going through `math_ops.abs` (which would add an unnecessary per-element sqrt).

The fallback also has to be safe under automatic differentiation. The QR kernel
(`fast=False`) has no custom backward pass, so routing it through
`control_flow_ops.cond` previously broke `tf.GradientTape` during graph
compilation (`Gradient not defined for fast=False`, and `tf.cond` is not
supported by gradient tape on graph control flow). This is fixed by wrapping the
fallback in `array_ops.stop_gradient` (so gradients never propagate up the
non-differentiable path) and branching by execution context: in eager mode the
fallback is taken directly from `factorization_failed.numpy()`, and in graph
mode `cond_v2.cond_v2` is used instead of `control_flow_ops.cond` because it
supports gradient tape.

**Verification**

- The reproducer above now returns `0.25` per row, matching `numpy.linalg.lstsq`
  exactly (max abs diff 0.0), instead of NaN.
- A singular 2x2 system returns `[0.2, 0.4]`, also matching `numpy.linalg.lstsq`.
- Identical results under `tf.function` (graph mode) and in eager mode.
- Well-conditioned inputs are unchanged: 24 full-rank cases (4 shapes x
  float32/float64 x 3 draws) produce bit-for-bit identical output before and
  after this change, so the extra work is one `reduce_any(is_nan(...))` and the
  fallback is never taken for them.
- Inputs that already contain NaN still produce NaN (the QR kernel returns NaN
  for them too), so no invalid input is silently "fixed".
- Adds `MatrixSolveLsOpTest.testSingularMatrix`, which runs in both graph and
  eager mode (correctly decorated so the neighboring `testBatchResultSize`
  retains its `use_gpu=True` decorator). `matrix_solve_ls_op_test.py` passes in
  full (101 tests).

Fixes #115505

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi there,

Thank you for your contribution! The logic accurately addresses the failing factorization. However, before this can be merged, there are a couple of required adjustments:

1. The `PyLint` check in the CI pipeline failed because `testSingularMatrix` has a duplicate `@test_util.run_in_graph_and_eager_modes` decorator.
2. For complex types, using `math_ops.abs` introduces unnecessary overhead since we just need to detect `NaN`. We can optimize the check by examining the real and imaginary parts separately.

Please apply the following patch to fix the lint error and include the optimization:

```diff
diff --git a/tensorflow/python/kernel_tests/linalg/matrix_solve_ls_op_test.py b/tensorflow/python/kernel_tests/linalg/matrix_solve_ls_op_test.py
index ca2cf9b21ab..3606fbf7639 100644
--- a/tensorflow/python/kernel_tests/linalg/matrix_solve_ls_op_test.py
+++ b/tensorflow/python/kernel_tests/linalg/matrix_solve_ls_op_test.py
@@ -170,7 +170,6 @@ class MatrixSolveLsOpTest(test.TestCase):
       self.assertEqual(tf_ans.shape, (2, 2))
 
   @test_util.run_in_graph_and_eager_modes(use_gpu=True)
-  @test_util.run_in_graph_and_eager_modes
   def testSingularMatrix(self):
     # The composite implementation is based on a Cholesky factorization of the
     # Gramian, which does not exist when the matrix is singular. It used to
diff --git a/tensorflow/python/ops/linalg_ops.py b/tensorflow/python/ops/linalg_ops.py
index d92631af0e9..a9b8fa66c3c 100644
--- a/tensorflow/python/ops/linalg_ops.py
+++ b/tensorflow/python/ops/linalg_ops.py
@@ -370,9 +370,13 @@ def _composite_impl(matrix, rhs, l2_regularizer):
       # here, and the fallback returns NaN for them as well, so there is no
       # need to inspect the inputs.
-      # `is_nan` is not defined for complex types, so reduce to the real
-      # magnitude first in that case.
-      failed_values = math_ops.abs(solution) if solution.dtype.is_complex else solution
-      factorization_failed = math_ops.reduce_any(math_ops.is_nan(failed_values))
+      # `is_nan` is not defined for complex types, so check the real and
+      # imaginary parts separately in that case.
+      if solution.dtype.is_complex:
+        factorization_failed = math_ops.reduce_any(math_ops.logical_or(
+            math_ops.is_nan(math_ops.real(solution)),
+            math_ops.is_nan(math_ops.imag(solution))))
+      else:
+        factorization_failed = math_ops.reduce_any(math_ops.is_nan(solution))
       return cond.cond(
           factorization_failed,
           lambda: gen_linalg_ops.matrix_solve_ls(
```

Once these changes are pushed, the tests will pass successfully. Thank you for fixing this edge case!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for the PR!

During our CI testing, we noticed a minor placement issue with the `testSingularMatrix` test in `matrix_solve_ls_op_test.py`. It appears you inserted the new test function directly *after* the `@test_util.run_in_graph_and_eager_modes(use_gpu=True)` decorator that was originally meant for `testBatchResultSize(self)`. 

This resulted in `testSingularMatrix` having duplicate decorators and `testBatchResultSize` missing its decorator entirely, which breaks the test suite.

Could you please move your new `testSingularMatrix` (along with its `@test_util.run_in_graph_and_eager_modes` decorator) to be *above* the `@test_util.run_in_graph_and_eager_modes(use_gpu=True)` decorator, so that `testBatchResultSize` correctly receives its decorator?

Thanks again!

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the detailed and helpful review! I've applied both suggestions:

1. Removed the duplicate `@test_util.run_in_graph_and_eager_modes` decorator on `testSingularMatrix`, and moved the test (with its decorator) above the `@test_util.run_in_graph_and_eager_modes(use_gpu=True)` decorator so that `testBatchResultSize` correctly receives its decorator again — this also fixes the test-suite breakage you spotted.
2. For complex dtypes, the NaN check now inspects the real and imaginary parts separately via `is_nan` instead of `math_ops.abs`, removing the unnecessary sqrt overhead.

Both changes are pushed in commit `d9f5ee2a`. Thanks again for catching these and for the clear patch — looking forward to your re-review.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for this PR! The functional behavior (returning valid results instead of NaNs for singular matrices) works perfectly, but the TensorFlow internal automated testing suite identified that the underlying change breaks `tf.GradientTape` backward passes. 

Specifically, the `linalg_grad_test` suite encounters this failure:
`NotImplementedError: tf.GradientTape.gradients() does not support graph control flow operations like tf.cond ...` and `ValueError: Gradient not defined for fast=False`.

Since the fallback QR solver operation `fast=False` does not have an implemented custom backward pass, inserting it inside a `cond` routing tree causes `tf.GradientTape` parsing to crash during graph compilation. We can mitigate this by stopping gradients from propagating up the fallback path and routing execution safely based on the execution context. 

Could you please apply the following modifications to `tensorflow/python/ops/linalg_ops.py` to fix the CI and pass internal presubmits?

```python
      # `is_nan` is not defined for complex types, so check the real and
      # imaginary parts separately in that case.
      if solution.dtype.is_complex:
        factorization_failed = math_ops.reduce_any(math_ops.logical_or(
            math_ops.is_nan(math_ops.real(solution)),
            math_ops.is_nan(math_ops.imag(solution))))
      else:
        factorization_failed = math_ops.reduce_any(math_ops.is_nan(solution))
        
      def fallback_fn():
        return array_ops.stop_gradient(
            gen_linalg_ops.matrix_solve_ls(
                matrix, rhs, l2_regularizer, fast=False))
                
      from tensorflow.python.eager import context
      if context.executing_eagerly():
        if factorization_failed.numpy():
          return fallback_fn()
        else:
          return solution
          
      from tensorflow.python.ops import cond_v2
      return cond_v2.cond_v2(
          factorization_failed,
          fallback_fn,
          lambda: solution)
```

Please let us know once this is updated!

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the additional catch — and for running the internal presubmit suites that surfaced the `tf.GradientTape` breakage. That was a gap I hadn't anticipated, so I really appreciate you identifying it.

I've pushed a fix in commit `03e8743d`:

1. The QR fallback (`fast=False`, which has no custom backward pass) is now wrapped in `array_ops.stop_gradient`, so gradients no longer try to propagate up the non-differentiable path — this resolves the `Gradient not defined for fast=False` error.
2. Instead of routing through `control_flow_ops.cond`, the fallback now branches by execution context:
   - In eager mode, it checks `factorization_failed.numpy()` directly and returns either the fallback or the original `solution`.
   - In graph mode, it uses `cond_v2.cond_v2`, which supports gradient tape, avoiding the `does not support graph control flow operations like tf.cond` failure.

The complex-type NaN check from the previous round is kept as-is. Could you please re-run the presubmits when you get a chance? Looking forward to your re-review, and thanks again for the thorough feedback — it made the change much more robust.

**Review by dmiltr3 — APPROVED** [review]
LGTM. 

I've successfully run the internal test suite against your latest changes addressing the Cholesky fallback logic. The fix works perfectly and passing tests confirm there's no breakage to `tf.GradientTape` graph control flows. 

Approving this PR. Thanks for the contribution!

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the thorough and patient review across all three iterations, and for approving this PR! I appreciate you running the internal presubmit suites each time. I will keep an eye out for the merge.

**Comment by nithyak0204** [comment]
Hi @Phoebus-Liu, Seems auto-merge is not happening but the changes are merged into master via commit [303300a](https://github.com/tensorflow/tensorflow/commit/303300abe50998de86b2acb09b3783f0502f3ed1), so we can close this PR.
Thank you for your contribution and helping improve TensorFlow!
